processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -0,0 +1,189 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment;
+
+import io.druid.segment.data.Indexed;
+
+
+/**
+ * Processing related interface
+ *
+ * A DimensionIndexer is a per-dimension stateful object that encapsulates type-specific operations and data structures
+ * used during the in-memory ingestion process (i.e., work done by IncrementalIndex).
+ *
+ * Ingested row values are passed to a DimensionIndexer, which will update its internal data structures such as
+ * a value->ID dictionary as row values are seen.
+ *
+ * The DimensionIndexer is also responsible for implementing various value lookup operations,
+ * such as conversion between an encoded value and its full representation. It maintains knowledge of the
+ * mappings between encoded values and actual values.
+ *
+ * NOTE:
+ * When encoding is present, there are two relevant orderings for the encoded values.
+ *
+ * 1.) Ordering based on encoded value's order of ingestion
+ * 2.) Ordering based on converted actual value
+ *
+ * Suppose we have a new String dimension DimA, which sees the values "Hello", "World", and "Apple", in that order.
+ * This would correspond to dictionary encodings of "Hello"=0, "World"=1, and "Apple"=2, by the order
+ * in which these values were first seen during ingestion.
+ *
+ * However, some use cases require the encodings to be sorted by their associated actual values.
+ * In this example, that ordering would be "Apple"=0, "Hello"=1, "World"=2.
+ *
+ * @param <EncodedType> class of the encoded values
+ * @param <ActualType> class of the actual values
+ *
+ */
+public interface DimensionIndexer<EncodedType extends Comparable<EncodedType>, ActualType extends Comparable<ActualType>>
+{
+
+  /**
+   * Given a list of row values (for multivalued dimensions) or single row value, update any
+   * internal data structures with the ingested values and return the row values as a Comparable[]
+   * to be used as a TimeAndDims key.
+   *
+   * The values within the Comparable[] return value should be encoded if applicable, i.e. as instances of EncodedType.
+   *
+   * NOTE: This function can change the internal state of the DimensionIndexer.
+   *
+   * @param dimValues Row value(s) to process
+   *
+   * @return A new DimensionIndexer object that synchronizes on the provided lock.
```

Comment:
It returns an `Object` not a `DimensionIndexer`.  If it returns a `DimensionIndexer` then the method signature should be `DimensionIndexer`.